chips/lowrisc/src/flash_ctrl.rs

```diff
@@ -0,0 +1,553 @@
+        };
+        self.registers.control.write(
+            CONTROL::OP::PROG
+                + CONTROL::PARTITION_SEL::INFO
+                + CONTROL::NUM.val(((PAGE_SIZE / 4) - 1) as u32)
+                + CONTROL::START::SET
+                + info_sel,
+        );
+
+        // Write the data until we are full or have written all the data
+        while !self.registers.status.is_set(STATUS::PROG_FULL)
+            && self.write_index.get() < buf.0.len()
+        {
+            let buf_offset = self.write_index.get();
+            let data: u32 = buf[buf_offset] as u32
+                | (buf[buf_offset + 1] as u32) << 8
+                | (buf[buf_offset + 2] as u32) << 16
+                | (buf[buf_offset + 3] as u32) << 24;
+
+            self.registers.prog_fifo.set(data);
+
+            self.write_index.set(buf_offset + 4);
+        }
+
+        // Save the buffer
+        self.write_buf.replace(buf);
+
+        // Enable interrupts and set the FIFO level
+        self.enable_interrupts();
+        self.registers.fifo_lvl.modify(FIFO_LVL::PROG.val(0xF));
+
+        Ok(())
+    }
+
+    fn erase_page(&self, page_number: usize) -> ReturnCode {
+        let addr = page_number * PAGE_SIZE;
+
+        if !self.data_configured.get() {
+            // If we aren't configured yet, configure now
+            self.configure_data_partition(self.region_num);
+        }
+
+        if !self.info_configured.get() {
+            // If we aren't configured yet, configure now
+            self.configure_info_partition(FlashBank::BANK0, self.region_num);
+        }
+
+        // Enable erase
+        self.registers
+            .mp_bank_cfg
+            .modify(MP_BANK_CFG::ERASE_EN_0::SET + MP_BANK_CFG::ERASE_EN_1::SET);
+
+        // Set the address
+        self.registers.addr.write(ADDR::START.val(addr as u32));
+
+        // Start the transaction
+        let info_sel = if self.region_num as u32 > 3 {
+            CONTROL::INFO_SEL::SET
+        } else {
+            CONTROL::INFO_SEL::CLEAR
+        };
+        self.registers.control.write(
+            CONTROL::OP::ERASE
+                + CONTROL::ERASE_SEL::PAGE
+                + CONTROL::PARTITION_SEL::INFO
+                + CONTROL::START::SET
+                + info_sel,
+        );
+
+        // Disable erase
+        self.registers
+            .mp_bank_cfg
+            .modify(MP_BANK_CFG::ERASE_EN_0::CLEAR + MP_BANK_CFG::ERASE_EN_1::CLEAR);
+
+        ReturnCode::SUCCESS
+    }
+}
+
+impl<'a> hil::nonvolatile_storage::NonvolatileStorage<'a> for FlashCtrl<'a> {
+    fn set_client(&self, client: &'a dyn hil::nonvolatile_storage::NonvolatileStorageClient<'a>) {
+        self.nvstorage_client.set(client);
+    }
+
+    fn read(&self, _buffer: &'a mut [u8], _address: usize, _length: usize) -> ReturnCode {
+        unimplemented!()
+    }
+
+    fn write(&self, _buffer: &'a mut [u8], _address: usize, _length: usize) -> ReturnCode {
+        unimplemented!()
+    }
+}
```

Comment:
I don't think it makes sense to provide this basically empty implementation of the `NonvolatileStorage` trait. Lots of chips (every chip?), implement only flash, and then use the nonvolatile storage capsule to provide the `NonvolatileStorage` interface, which looks like what you want to do here as well.

Reply:
Removed.